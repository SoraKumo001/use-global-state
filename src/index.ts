import { Dispatch, SetStateAction, useEffect, useState } from 'react';
type Render = Dispatch<SetStateAction<{}>>;
const renderMap = new Map<string, Set<Render>>();
const initialKyes = new Set<string>();
const cache: { [key: string]: unknown } = {};

export const useGlobalState = <T>(key: string, initialData?: T) => {
  const [, render] = useState<{} | null>(null);
  const dispatch = (data?: T | ((data: T) => T)) => {
    if (data === undefined) delete cache[key];
    else if (typeof data === 'function') {
      cache[key] = (data as (data: T) => T)(cache[key] as T);
    } else cache[key] = data;
    Array.from(renderMap.get(key) || []).forEach((render) => render({}));
  };
  useEffect(
    () => () => {
      renderMap.get(key)?.delete(render);
    },
    []
  );
  if (initialData !== undefined && !initialKyes.has(key)) {
    initialKyes.add(key);
    cache[key] = initialData;
  }
  if (renderMap.has(key)) {
    renderMap.get(key)?.add(render);
  } else {
    const e = new Set<Render>();
    e.add(render);
    renderMap.set(key, e);
  }
  return [cache[key] as undefined | T, dispatch] as const;
};

export const mutate = async <T = Object>(
  key?: string,
  data?: T | Promise<T> | ((data: T) => T | Promise<T>)
) => {
  if (key) {
    if (data !== undefined) {
      if (typeof data === 'function')
        cache[key] = await (data as (data: T) => T | Promise<T>)(cache[key] as T);
      else cache[key] = await data;
    }
    renderMap.get(key)?.forEach((render) => render({}));
  } else {
    renderMap.forEach((renders) => renders.forEach((render) => render({})));
  }
};
