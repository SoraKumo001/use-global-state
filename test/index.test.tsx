import React, { useEffect } from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import { reset, useGlobalState } from '../src/index';

let container: HTMLElement;
beforeEach(() => {
  reset();
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
});

it('initData undefined', () => {
  const Component01 = () => {
    const [value] = useGlobalState('Key1');
    return <>{value ?? 'undefined'}</>;
  };
  const Component02 = () => {
    const [value] = useGlobalState('Key1');
    return <>{value ?? 'undefined'}</>;
  };
  const Component03 = () => {
    const [value] = useGlobalState('Key1');
    return <>{value ?? 'undefined'}</>;
  };

  act(() => {
    render(
      <>
        <Component01 />
        <Component02 />
        <Component03 />
      </>,
      container
    );
  });
  expect(container.childNodes).toMatchSnapshot();
});

it('initData', () => {
  const Component01 = () => {
    const [value] = useGlobalState('Key1');
    return <>{value ?? 'undefined'}</>;
  };
  const Component02 = () => {
    //first initData
    const [value] = useGlobalState('Key1', 1);
    return <>{value ?? 'undefined'}</>;
  };
  const Component03 = () => {
    const [value] = useGlobalState('Key1', 2);
    return <>{value ?? 'undefined'}</>;
  };

  act(() => {
    render(
      <>
        <Component01 />
        <Component02 />
        <Component03 />
      </>,
      container
    );
  });
  expect(container.childNodes).toMatchSnapshot();
});

it('useEffect', () => {
  const Component01 = () => {
    const [value] = useGlobalState('Key1');
    return <>{value ?? 'undefined'}</>;
  };
  const Component02 = () => {
    const [value, setValue] = useGlobalState('Key1', 1);
    useEffect(() => {
      setValue(3);
    }, []);
    return <>{value ?? 'undefined'}</>;
  };
  const Component03 = () => {
    const [value, setValue] = useGlobalState<unknown>('Key1', 2);
    useEffect(() => {
      //Last setValue
      setValue('Effect');
    }, []);
    return <>{value ?? 'undefined'}</>;
  };

  act(() => {
    render(
      <>
        <Component01 />
        <Component02 />
        <Component03 />
      </>,
      container
    );
  });
  expect(container.childNodes).toMatchSnapshot();
});

it('onclick', () => {
  const Component01 = () => {
    const [value, setValue] = useGlobalState('Key1');
    return (
      <button
        id="click"
        onClick={() => {
          setValue('Click');
        }}
      >
        {value}
      </button>
    );
  };
  const Component02 = () => {
    const [value, setValue] = useGlobalState<unknown>('Key1', 1);
    useEffect(() => {
      setValue('Efect');
    }, []);
    return <>{value ?? 'undefined'}</>;
  };
  const Component03 = () => {
    const [value] = useGlobalState('Key1', 2);
    return <>{value ?? 'undefined'}</>;
  };

  act(() => {
    render(
      <>
        <Component01 />
        <Component02 />
        <Component03 />
      </>,
      container
    );
  });
  expect(container.childNodes).toMatchSnapshot();
  act(() => {
    const element = container.querySelector('#click');
    element?.dispatchEvent(new MouseEvent('click', { bubbles: true }));
  });
  expect(container.childNodes).toMatchSnapshot();
});
