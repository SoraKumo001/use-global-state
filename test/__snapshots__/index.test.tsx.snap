// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initData 1`] = `
NodeList [
  <div>
    1
  </div>,
  <div>
    1
  </div>,
  <div>
    1
  </div>,
]
`;

exports[`initData undefined 1`] = `
NodeList [
  <div />,
  <div />,
  <div />,
]
`;

exports[`onclick 1`] = `
NodeList [
  <button
    id="click"
  >
    3
  </button>,
  <div>
    3
  </div>,
  3,
]
`;

exports[`onclick 2`] = `
NodeList [
  <button
    id="click"
  >
    4
  </button>,
  <div>
    4
  </div>,
  4,
]
`;

exports[`useEffect 1`] = `
NodeList [
  <div>
    4
  </div>,
  <div>
    4
  </div>,
  4,
]
`;
